### PR TITLE
Unify mixers by basing channel count on module params

### PR DESCRIPTION
**Pull request by bjeanes**
This also introduces an 8-channel mixer.

image

The eventual idea being that we could have a amixer that can add/remove channels dynamically, if needed.

A screenshot showing mixing on three channels and cue-ing out the 8th channel:

WhatsApp Image 2020-04-09 at 4 03 00 PM

### CODE REVIEW

**Comment by bjeanes** [comment]
You may have a reason not to unify this in which case: no problem, I'm not attached to this!

I did this mostly as an excuse to wrestle with some more Rust :)

File: frontend/src/workspace.rs [hunk 1]
```diff
@@ -492,10 +492,13 @@ impl Workspace {
             _ => return html! {},
         };
 
+        crate::log!("{:?}", MixerParams::default());
```
Comment (haileys) [inline frontend/src/workspace.rs:495]: 👀

File: protocol/src/lib.rs [hunk 2]
```diff
@@ -201,9 +199,23 @@ impl Default for EnvelopeParams {
...
+    pub fn with_channels(n: usize) -> MixerParams {
+        MixerParams {
+            channels: vec![MixerChannelParams::default(); n]
+        }
+    }
+}
+
+impl Default for MixerParams {
+    fn default() -> Self {
+        Self::with_channels(4)
```
Comment (haileys) [inline protocol/src/lib.rs:217]: Does `MixerParams` still need to impl `Default` now? I'm ok with a consequence of this refactor being that there's no default mixer anymore

File: src/module/mixer.rs [hunk 3]
```diff
@@ -1,54 +1,61 @@
...
     }
 
     fn run_tick(&mut self, _t: u64, inputs: &[Option<&[Sample]>], outputs: &mut [&mut [Sample]]) -> Option<Self::Indication> {
+        const MASTER: usize = 0;
+        const CUE: usize = 1;
+
         let len = outputs[0].len();
 
-        let mut channel_gain: [f64; 4] = [0.0; 4];
+        let mut channel_gain = vec![0.0; self.params.channels.len()];
```
Comment (haileys) [inline src/module/mixer.rs:41]: Can we move this allocation into the state struct?

File: src/module/mixer.rs [hunk 4]
```diff
@@ -1,54 +1,61 @@
...
 
     fn update(&mut self, params: Self::Params) -> Option<Self::Indication> {
+        self.inputs = params.channels.iter().map(|_| LineType::Stereo ).collect();
         self.params = params;
         None
     }
 
     fn run_tick(&mut self, _t: u64, inputs: &[Option<&[Sample]>], outputs: &mut [&mut [Sample]]) -> Option<Self::Indication> {
+        const MASTER: usize = 0;
+        const CUE: usize = 1;
```
Comment (haileys) [inline src/module/mixer.rs:37]: 👍

**Review by haileys — COMMENTED** [review]
This looks great. Agree with this direction. Left a couple comments below:

File: protocol/src/lib.rs (hunk 2 above)
Comment (bjeanes) [reply protocol/src/lib.rs:217]: Yeah seems reasonable. I can pull it.

Comment (bjeanes) [reply protocol/src/lib.rs:217]: Ah I remember my reason. It's just so `MixerParams` _has_ a default. It previously derived `Default` so I just wanted to preserve that interface. But yeah, I have no problem removing it either.

**Comment by bjeanes** [comment]
I deleted my fork forgetting this PR was tied to it. I recreated fork but the link is broken so I can't update this PR. However, I'm going to address all those items and, since they are all quite minor, do the merge locally and push it to master.

File: frontend/src/workspace.rs (hunk 1 above)
Comment (bjeanes) [reply frontend/src/workspace.rs:495]: Fixed in 69184eb

File: protocol/src/lib.rs (hunk 2 above)
Comment (bjeanes) [reply protocol/src/lib.rs:217]: Fixed in 32f240a

File: src/module/mixer.rs (hunk 3 above)
Comment (bjeanes) [reply src/module/mixer.rs:41]: Fixed in e33fa07

**Pull request merged**